### ISSUE TITLE
useless-super-delegation incorrect warning

### ISSUE THREAD

**Issue by pjljvandelaar** [PyCQA/pylint#3934]
### Steps to reproduce
1. With pylint
### pylint --version output
```
pylint 2.6.0
astroid 2.4.2
Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)]
```
on windows 10

2. Given the following file:
```
class Variable(Reference):
    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Variable):
            return False

        return super().__eq__(other)

    def __hash__(self) -> int:
        return super().__hash__()
```

3. `pylint variable.py` results in:
```
variable.py:18:4: W0235: Useless super delegation in method '__hash__' (useless-super-delegation)
```
4. Yet removing this `__hash__` function and running python 3.9.0 results in
```
TypeError: unhashable type: 'Variable'
```
whenever it is used in a hashable collection (such as a set).

### Current behavior
I don't know how correct the python code is, but
either python or pylint is wrong.

### Expected behavior
Make python and pylint agree: either by removing, changing, or generating another warning.

**Comment by dbaty** [comment]
Could you provide more details about the `Reference` base class that you use? Does it define a `__hash__()` method? Or does it subclass a class that defines such a method?

**Comment by pjljvandelaar** [comment]
Dear dbaty,

I didn't though the exact structure would be relevant.
But now you asked:
The example is a little more complicated than you thought (but it was not the first time I experienced this warning).

Reference is an Abstract Base Class inheriting from Expression.
Expression is an Abstract Base Class.
Multiple instance of Reference are made, such as a Variable.

The code is below
```
from abc import ABC
from dataclasses import dataclass

from expression import Expression
from file_path import FilePath
from file_range import FileRange
from name import Name


@dataclass(frozen=True)
class Reference(Expression, ABC):
    """
    Since nested functions in Ada can refer to a hidden homonym in an enclosing scope
    by using the name of the enclosing scope as a prefix,
    we need similar functionality.
    We realize the distinction of references by adding file path and file range to the name.
    """
    name: Name
    file_path: FilePath
    file_range: FileRange

    def __repr__(self) -> str:
        return '[' + str(self.name) + ']'

    def is_closed(self) -> bool:
        return False

    def contains_expression(self, expression) -> bool:
        return self == expression
```

Expression is 
```
from abc import ABC, abstractmethod


class Expression(ABC):
    @abstractmethod
    def is_closed(self) -> bool:
        """
        Is the expression closed?
        I.e. are no free variables present?
        """

    @abstractmethod
    def contains_expression(self, expression) -> bool:
        """
        Does this expression contain the given expression?
        """
```
and a reference example: the class Variable
```
from reference import Reference


class Variable(Reference):
    """
    Since nested functions in Ada can refer to a hidden homonym in an enclosing scope
    by using the name of the enclosing scope as a prefix,
    we need similar functionality.
    We realize the distinction of variables by adding file path and file range to the name.
    """

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Variable):
            return False

        return super().__eq__(other)

    def __hash__(self) -> int:
        return super().__hash__()

    def is_closed(self) -> bool:
        return False
```

**Comment by dbaty** [comment]
I can reproduce what you report. (See below for a minimal example, where I removed the dataclass decorator on `Reference` since it does not seem to be part of the issue.)



```python
# pylint: disable=missing-module-docstring, missing-class-docstring
# pylint: disable=too-few-public-methods

from abc import ABC


class Expression(ABC):
    pass


class Reference(Expression, ABC):
    pass


class Variable(Reference):
    def __eq__(self, other):
        if not isinstance(other, Variable):
            return False
        return super().__eq__(other)


set([Variable()])
```

**Comment by dbaty** [comment]
I can reproduce what you report. See a minimal example at the end of my comment. (I removed everything that was not necessary to reproduce the issue. If you could try to do so the next time your report an issue, that would be great! Thank you! :) )

I think that both Python and pylint are right here:

1. Python is right. As per my understanding of [the documentation of \_\_hash__()](https://docs.python.org/3/reference/datamodel.html#object.__hash__), your use case requires you to define `__hash__()`:

    > A class that overrides __eq__() and does not define __hash__() will have its __hash__() implicitly set to None. When the __hash__() method of a class is None, instances of the class will raise an appropriate TypeError when a program attempts to retrieve their hash value, and will also be correctly identified as unhashable [...]

2. On the other hand, pylint frowns upon this kind of constructs:

    ```python
    class C:
        def foo(self):
            return super().foo()
    ```

    ... and reports `useless-super-delegation` because it's indeed useless.

It looks like overriding `__hash__()` like that is a special case that could be allowed by pylint. The check happens in [ClassChecker._check_useless_super_delegation()](https://github.com/PyCQA/pylint/blob/b725f8a94c809900e945516ae26c956318a00d6c/pylint/checkers/classes.py#L1012). It could be a good first contribution. Would you be willing to have a go at it? :)


```python
# pylint: disable=missing-module-docstring, missing-class-docstring
# pylint: disable=too-few-public-methods

from abc import ABC


class Expression(ABC):
    pass


class Variable(Expression):
    def __eq__(self, other):
        if not isinstance(other, Variable):
            return False

        return super().__eq__(other)

    def __hash__(self):
        return super().__hash__()


a = Variable()
b = Variable()
assert len({a, a, b}) == 2
```

**Comment by pjljvandelaar** [comment]
Looked at the code, but I wouldn't know how to add the desired functionality.
Seems not just adding at the beginning
```
if function.name == "__hash__"
  return 
```
to me.

Furthermore, don't know what unit test should be added (and how- pytest/unittest/ own test framework)...

**Comment by hippo91** [comment]
@pjljvandelaar the easiest way to check your implementation is to add a method in the unittest: `useless_super_delegation.py` (or one similar) in the `tests/functional/u` directory. Do not forget to add the awaited result in the `useless_super_delegation.txt` file in the same directory.
You can launch the test suite with `python -m tox -epy38` (for python 3.8, -epy37 for python 3.7 etc...). You can find documentation in the README file.
